### ISSUE TITLE
Documentation and internal symbol exports

### ISSUE THREAD

**Issue by phoe** [pcostanza/closer-mop#12]
Multiple exports from this system are undocumented. Some of them need to be fixed elsewhere, for example `CLASSP` on SBCL, but some are internal to Closer to MOP, for example, `FIX-INITARGS`.

Another issue with it is that Closer to MOP reexports an internal SBCL symbol, which is against SBCL guidelines - that symbol may disappear at any moment, breaking Closer to MOP.

**Comment by pcostanza** [comment]
1) I'm not sure what you mean with "undocumented." There is documentation in README.md and features.txt. Are you asking for :documentation strings? What problem would that solve?

2) Closer to MOP reexports internal symbols of underlying Common Lisp implementations if they happen to be named and behave exactly like specified in AMOP. I would then consider it a bug of the underlying implementation not to export those.

Again, it's not clear to me what problem would be solved. Closer to MOP is already quite old by now, and this has not caused any issues so far. I would prefer to focus on actual problems.

Do you expect any such symbols to disappear any time soon?

**Comment by phoe** [comment]
1. Yes - I mean a lack of documentation strings. Some programmers use `DESCRIBE` to display documentation strings for functions during programming.

2. Yes, I think this should be discussed with SBCL. The only issue I see is that `CLASSP` seems to be defined by Closer to MOP and not [MOP itself](http://metamodular.com/CLOS-MOP/all-no-methods.html); if such is the case, I would expect the symbol to have a documentation string.

3. Recently ironclad, nibbles, and weblocks were affected by SBCL removing a few internal symbols from the `sb-vm` package. The disappearance of those symbols caused these packages to stop building. The general consensus I have heard

I am willing to prepare and submit the patches that add the missing documentation strings.

**Comment by Hexstream** [comment]
For what it's worth, as someone who loves good documentation and hates docstrings, I think it's reasonable to expect good documentation and unreasonable to expect docstrings.

-------------------

`The general consensus I have heard`

... Yes?

(Also, I'm not too familiar with SBCL internals but my impression is that `sb-vm` is a fairly "internal" package compared to, say, `sb-ext`. So I'm not sure I would tend to generalize lessons from `sb-vm` instability...)

**Comment by phoe** [comment]
Welp, I forgot to type that part.

The general consensus I have heard in `#sbcl` on Freenode is that if an external system uses symbols that are either internal in a SBCL-defined package (think the aforementioned `sb-pcl::classp`) or in internal SBCL packages (think `sb-vm`), there are two ways of resolving that conflict: either by changing the external system to stop using the internal symbol, or by documenting and exporting the symbol as an external API defined by SBCL. The option where an internal symbol is used is undesired.

**Comment by pcostanza** [comment]
It's important to keep the history of Closer to MOP in mind: I started the project about 14 years ago because back then, the quality of CLOS MOP adherence across several Common Lisp implementations was pretty bad. I wanted to make sure that initially AspectL and then ContextL would be portable across different Common Lisp implementations, but because they are pushing the CLOS MOP quite hard, a lot of fixes and workarounds were necessary to do so. I then considered it to be cleaner to separate out these fixes into what became Closer to MOP.

Ideally, Closer to MOP should not be necessary. Despite its flaws, the CLOS MOP specification in AMOP is pretty good, and apart from selecting different package names, there should be no need to have such a compatibility layer. Closer to MOP was supposed to be a temporary solution, and my - admittedly naive - hope was that it would someday become unnecessary.

I haven't completely lost that hope yet, although it seems quite difficult to achieve (especially when it comes to the commercial implementations).

That's why I'm hesitating to invest time into things that don't contribute to that actual goal... ;)

**Comment by phoe** [comment]
Well, nowadays, Closer to MOP [is the eigth most popular package on Quicklisp](https://www.quicklisp.org/tmp/stats.txt) and the de facto portability library for using MOP. If someone wants to use MOP functionalities, they rarely rely on implementation-defined libraries; they usually use Closer to MOP instead.

So, if your goal was to eventually reduce Closer to MOP's role in the Lisp world, you have spectacularly failed at it, and I actually applaud you for that - instead of multiple implementation-defined MOP packages, we have one package with consistent behaviour that we need to care about in order to be able to use MOP in our code, no matter which Lisp implementation we run our code on.

**Comment by Hexstream** [comment]
Actually, that's *quite a bit* of an understatement.
(The undated Quicklisp download stats file you linked to seems a bit misleading in context.)

According to *all* official Quicklisp download stats published in 2017 and 2018, closer-mop was consistently #3 or better (always #2 except for one month):

[#2 in June 2017](http://blog.quicklisp.org/2017/07/june-2017-quicklisp-download-stats.html)
[#3 in July 2017](http://blog.quicklisp.org/2017/08/july-2017-quicklisp-download-stats.html)
[#2 in October 2017](http://blog.quicklisp.org/2017/11/october-2017-download-stats.html)
[#2 in November 2017](http://blog.quicklisp.org/2017/12/november-2017-quicklisp-download-stats.html)
[#2 in December 2017](http://blog.quicklisp.org/2018/01/download-stats-for-december-2017.html)
[#2 in January 2018](http://blog.quicklisp.org/2018/02/download-stats-for-january-2018.html)
[#2 in February 2018](http://blog.quicklisp.org/2018/03/download-stats-for-february-2018.html)

(Unfortunately, there were not many published Quicklisp download stats in 2018...)

Additionally, quoting myself from another one of my [comments](https://github.com/robert-strandh/CLOS-MOP-HTML/issues/4#issuecomment-346235403) from 21 november 2017 (the situation hasn't substantially changed since):

> 2. According to [this](http://quickdocs.org/closer-mop/), closer-mop has 97 reverse dependencies in Quicklisp. From what I understand of [this](https://github.com/quicklisp/quicklisp-projects/tree/master/projects), there are currently 1608 projects in Quicklisp. So 6% of libraries in Quicklisp DIRECTLY depend on the MOP. That may not sound like much to some, but as far as I'm aware, VERY few Common Lisp libraries have anywhere close to this many direct reverse dependencies. (I haven't bothered checking what the number of indirect dependencies is, I'm sure it's very high, *if someone could provide that number it would be great!*)

Additionally, the MOP is quite often mentioned in Freenode #lisp, as I [mentioned](https://twitter.com/HexstreamSoft/status/928515550527000576) a few times. This should be thoroughly unsurprising to anyone who has a clue, given the obvious importance of the MOP, which [surprisingly and unfortunately does not seem to include Rainer Joswig](https://twitter.com/HexstreamSoft/status/1047703410944958464). I think this illustrates well how important and criminally underrated the CLOS MOP is, and by extension, closer-mop. Oh and I didn't even mention here that "somehow" almost nobody gave a shit about [this](https://clos-mop.hexstreamsoft.com/), but obviously it would be pretty hard for me to recount any details about it without getting rude...

Anyway, in a nutshell, closer-mop is one of the most important and most underrated libraries in the Common Lisp ecosystem.

The CLOS MOP is quite literally [one of our crown jewels](https://twitter.com/HexstreamSoft/statuses/1019718221514108929), and I wish more people understood and acknowledged this.

**Comment by pcostanza** [comment]
Hi Michał,

Thanks a lot for the clarifications, especially the one you sent privately. I'm happy to accept patches, but there are some caveats:

- Patches must work equally well for all supported Common Lisp implementations. Correct me if I'm wrong, but it sounds a bit like you primarily care about SBCL. However, Closer to MOP has to provide the same, or at least similar functionality and features for the other Common Lisps as well. I'm willing to spend some amount of time myself to ensure that this is the case, but I'd prefer patches to be as complete as possible in this regard.

- I'm skeptical that I will accept patches that try to find different solutions than to export implementation-dependent internal symbols. The CLOS MOP emphasizes the importance of protocols, so the focus cannot be only that single functions work correctly, but that the interplay between families of functions is also correct. You can use the separate repo mop-features, and the test suites for aspectl, contextl and filtered-functions as good approximations for ensuring this.

- My time to work on Closer to MOP is limited these days, so I may react only very slowly.

**Comment by phoe** [comment]
Hey Pascal,

I use SBCL on all development machines of mine, that's correct. But, as a part of the work I want to do on the library, I am willing to ensure that the patches work equally well across all implementations - meaning that, e.g., if some symbols are exported from internal packages, I will try to work with the maintainers of these implementations (wherever possible) to export them properly.

I'll also try to make these patches as easy to digest and handle as I can in order to limit the time you'll need to spend on them.
